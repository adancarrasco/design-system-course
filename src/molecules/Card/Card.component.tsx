import * as React from 'react'
import styled from '@emotion/styled'
import { ChatBubbleOutlineOutlined, EditOutlined, VisibilityOutlined } from '@material-ui/icons'

import Tag, { ITagProps } from '../../atoms/Tag'
import Badge, { IBadgeProps } from '../../atoms/Badge'
import { Typography } from '../../atoms'
// TODO:
// Think of adding DueDate

export interface ICardProps extends ICardStyledProps {
  children: React.ReactNode | string
  tagsProps?: ITagProps[]
  isWatching?: boolean
  commentsCount?: number
  watchers?: IBadgeProps[]
}

const Card: React.FC<ICardProps> = ({
  children,
  tagsProps,
  isWatching,
  commentsCount,
  watchers,
  ...rest
}) => (
  <CardStyled role="button" {...rest}>
    <EditOutlinedStyled fontSize="small" />
    <TagsContainerStyled>
      {tagsProps && tagsProps.map((tagProps, i) => <TagStyled key={i} {...tagProps} />)}
    </TagsContainerStyled>
    {children}
    <StatusContainerStyled>
      <div>
        {isWatching && <VisibilityOutlined fontSize="small" />}
        {commentsCount && (
          <>
            <ChatBubbleOutlineOutlined fontSize="small" />
            <Typography>{commentsCount}</Typography>
          </>
        )}
      </div>
      <div>
        {watchers &&
          watchers.map(({ userName }) => <Badge size="small" key={userName} userName={userName} />)}
      </div>
    </StatusContainerStyled>
  </CardStyled>
)

const EditOutlinedStyled = styled(EditOutlined)`
  &.MuiSvgIcon-root {
    display: none;
    position: absolute;
    right: 5px;
    top: 5px;
  }
`

export interface ICardStyledProps extends IStatusContainerStyledProps {
  backgroundColor?: string
  backgroundColorHover?: string
  boxShadow?: string
}
const CardStyled = styled.div<ICardStyledProps>`
  background-color: ${(props) => props.backgroundColor ?? '#fff'};
  box-shadow: ${(props) => props.boxShadow ?? '0 1px 0 rgba(9,30,66,.25)'};
  box-sizing: border-box;
  color: ${(props) => props.color};
  cursor: pointer;
  display: flex;
  flex-direction: column;
  padding: 0.4rem;
  position: relative;
  width: 100%;
  :hover {
    .MuiSvgIcon-root:first-of-type {
      display: inline-flex;
    }
    background-color: ${(props) => props.backgroundColorHover ?? '#f4f5f6'};
  }
`

export interface ITagsContainerStyledProps {}
const TagsContainerStyled = styled.div<ITagsContainerStyledProps>`
  display: flex;
  margin-bottom: 0.3rem;
`

export interface IStatusContainerStyledProps {
  color?: string
}
const StatusContainerStyled = styled.div<IStatusContainerStyledProps>`
  color: ${(props) => props.color};
  margin-top: 0.3rem;
  display: flex;
  justify-content: space-between;
  & > div:first-of-type {
    color: #ababab;
    & > svg {
      margin-right: 0.5rem;
      vertical-align: middle;
    }
  }
`
const TagStyled = styled(Tag)`
  margin-right: 0.4rem;
`

export default Card
