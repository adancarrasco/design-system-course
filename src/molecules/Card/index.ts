export { default } from './Card.component'
export type { ICardProps } from './Card.component'
