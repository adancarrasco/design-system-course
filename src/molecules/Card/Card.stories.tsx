import React from 'react'
import { storiesOf } from '@storybook/react'

import Card from '.'

storiesOf('Molecules|Card', module)
  .add('Default', () => <Card>Test</Card>)
  .add('With status props', () => (
    <Card
      tagsProps={[{ backgroundColor: 'red' }, { backgroundColor: 'blue' }]}
      commentsCount={3}
      isWatching
      watchers={[{ userName: 'Adan Carrasco' }]}
    >
      Test
    </Card>
  ))
