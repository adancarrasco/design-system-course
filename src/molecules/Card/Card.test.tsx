import React from 'react'
import { render } from '@testing-library/react'
import { matchers } from 'jest-emotion'
import '@testing-library/jest-dom/extend-expect'

import Card, { ICardProps } from '.'

expect.extend(matchers)

describe('Card Component', () => {
  const renderComponent = (props: ICardProps, dataTestId: string) =>
    render(
      <Card data-testid={dataTestId} {...props}>
        {props.children}
      </Card>,
    )

  it('matches snapshot and default render', () => {
    const dataTestId = 'card'
    const props = {
      children: 'Test',
      tagsProps: [{ backgroundColor: 'red' }],
      commentsCount: 2,
    }
    const { getByTestId } = renderComponent(props, dataTestId)
    const container = getByTestId('card')
    expect(container).toMatchSnapshot()
    expect(container).toBeInTheDocument()
  })

  it('should render with custom props', () => {
    const dataTestId = 'card'
    const props = {
      children: 'Test',
      tagsProps: [{ backgroundColor: 'red' }],
      commentsCount: 2,
      watchers: [{ userName: 'Adan Carrasco' }],
    }
    const { getByTestId, getByText, debug } = renderComponent(props, dataTestId)
    const container = getByTestId('card')
    expect(container).toBeInTheDocument()
    const badgeContainer = getByText('AC')
    const commentsCountContainer = getByText('2')
    expect(badgeContainer).toBeInTheDocument()
    expect(commentsCountContainer).toBeInTheDocument()
  })
})
