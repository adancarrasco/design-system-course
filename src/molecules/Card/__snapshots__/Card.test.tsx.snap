// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component matches snapshot and default render 1`] = `
.emotion-5 {
  background-color: #fff;
  box-shadow: 0 1px 0 rgba(9,30,66,.25);
  box-sizing: border-box;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0.4rem;
  position: relative;
  width: 100%;
}

.emotion-5:hover {
  background-color: #f4f5f6;
}

.emotion-5:hover .MuiSvgIcon-root:first-of-type {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0.MuiSvgIcon-root {
  display: none;
  position: absolute;
  right: 5px;
  top: 5px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 0.3rem;
}

.emotion-1 {
  background-color: red;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 3rem;
  height: 0.4rem;
  border-radius: 3px;
  margin-right: 0.4rem;
}

.emotion-1:hover {
  background-color: #ebebeb;
}

.emotion-4 {
  margin-top: 0.3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-4 > div:first-of-type {
  color: #ababab;
}

.emotion-4 > div:first-of-type > svg {
  margin-right: 0.5rem;
  vertical-align: middle;
}

<div
  class="emotion-5"
  data-testid="card"
  role="button"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root emotion-0 MuiSvgIcon-fontSizeSmall"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M14.06 9.02l.92.92L5.92 19H5v-.92l9.06-9.06M17.66 3c-.25 0-.51.1-.7.29l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.2-.2-.45-.29-.71-.29zm-3.6 3.19L3 17.25V21h3.75L17.81 9.94l-3.75-3.75z"
    />
  </svg>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-1"
    />
  </div>
  Test
  <div
    class="emotion-4"
  >
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 2H4c-1.1 0-2 .9-2 2v18l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm0 14H6l-2 2V4h16v12z"
        />
      </svg>
      <span
        class="emotion-3"
      >
        2
      </span>
    </div>
    <div />
  </div>
</div>
`;
